docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,282 @@
+| `RESTRICTED_USER` | A special privilege to say that their access can’t be changed by `SUPER` users. Statements `DROP USER`, `SET PASSWORD`, `ALTER USER`, `REVOKE` are all limited. | It is intended for the CloudAdmin user in DbaaS. |
+| `RESTRICTED_TABLES` | A special privilege which means that the SEM hidden table semantic doesn’t apply. | It is intended for the CloudAdmin user in DbaaS. | 
+
+### Parser Changes
+
+The parser already supports `DYNAMIC` privileges. This can be confirmed by the following patch to TiDB, where they are sent as the static type of `ExtendedPriv`:
+
+```
+diff --git a/planner/core/planbuilder.go b/planner/core/planbuilder.go
+index 90d5b9e82..d1644ee83 100644
+--- a/planner/core/planbuilder.go
++++ b/planner/core/planbuilder.go
+@@ -2298,6 +2298,11 @@ func collectVisitInfoFromGrantStmt(sctx sessionctx.Context, vi []visitInfo, stmt
+ 
+        var allPrivs []mysql.PrivilegeType
+        for _, item := range stmt.Privs {
++
++               if item.Priv == mysql.ExtendedPriv {
++                       fmt.Printf("### Attempting to set DYNAMIC privilege: %s\n", item.Name)
++               }
++
+                if item.Priv == mysql.AllPriv {
+                        switch stmt.Level.Level {
+                        case ast.GrantLevelGlobal:
+```
+
+This results in the following written to the log file:
+```
+mysql> grant acdc on *.* to u1;
+ERROR 8121 (HY000): privilege check fail
+
+### Attempting to set DYNAMIC privilege: acdc
+```
+
+It might be possible that changes are still required if `ExtendedPriv` is not supported in all contexts (REVOKE, etc).
+
+Note that creating a role with the same name as a DYNAMIC privilege is supported. A `GRANT` statement can be attributed to ROLES when it omits the ON *.* syntax. Thus:
+
+```
+GRANT BINLOG_ADMIN TO u1; // grants the role binlog_admin
+GRANT BINLOG_ADMIN ON *.* TO u1; // grants the dynamic privilege binlog_admin
+```
+
+This same nuance applies to MySQL.
+
+### Documentation Plan
+
+The statement reference pages for each of the affected metadata commands will need to be updated to describe dynamic privileges.
+
+There will also need to be documentation specific to `DYNAMIC` privileges to describe how it works, and the purpose of fine-grained access control.
+
+
+## Test Design
+
+Testing dynamic privileges is a little bit complex because of the various ways privileges can be inherited by users:
+
+* Direct `GRANT` to the user
+* Granting to a role that the user inherits.
+
+### Functional Tests
+
+Unit tests will be added to cover the semantics around role/dynamic privilege precedence, including logical restoring in a different order.
+
+Integration testing needs to test with global kill enabled/disabled.
+
+### Scenario Tests
+
+The use-cases required by the DBaaS team should be validated when combined with `security-enhanced-mode`. They are:
```

Comment:
We also need to add scenario tests for:

- all the dynamic privileges
- several user-defined dynamic privileges

By validating what they can and can't do to ensure the dynamic privilege works as expected.

Reply:
OK, I've added it.